crates/core/app/tests/app_can_undelegate_from_a_validator.rs

```diff
@@ -0,0 +1,374 @@
+                memo::MemoPlaintext, plan::MemoPlan, TransactionParameters, TransactionPlan,
+            },
+        };
+        let snapshot = storage.latest_snapshot();
+        client.sync_to_latest(snapshot.clone()).await?;
+        let rate = snapshot
+            .get_validator_rate(&identity_key)
+            .await?
+            .ok_or(anyhow::anyhow!("new validator has a rate"))?
+            .tap(|rate| tracing::info!(?rate, "got new validator rate"));
+
+        let undelegation_id = DelegationToken::new(identity_key).id();
+        let note = client
+            .notes
+            .values()
+            .filter(|n| n.asset_id() == undelegation_id)
+            .cloned()
+            .next()
+            .expect("the test account should have one staking token note");
+        let spend = SpendPlan::new(
+            &mut rand_core::OsRng,
+            note.clone(),
+            client
+                .position(note.commit())
+                .expect("note should be in mock client's tree"),
+        );
+        let undelegate = rate.build_undelegate(
+            storage.latest_snapshot().get_current_epoch().await?,
+            note.amount(),
+        );
+        let undelegate_token_id = undelegate.unbonding_token().id();
+        let output = OutputPlan::new(
+            &mut rand_core::OsRng,
+            undelegate.unbonded_value(),
+            *test_keys::ADDRESS_1,
+        );
+
+        let mut plan = TransactionPlan {
+            actions: vec![spend.into(), output.into(), undelegate.into()],
+            // Now fill out the remaining parts of the transaction needed for verification:
+            memo: MemoPlan::new(&mut OsRng, MemoPlaintext::blank_memo(*test_keys::ADDRESS_0))
+                .map(Some)?,
+            detection_data: None, // We'll set this automatically below
+            transaction_parameters: TransactionParameters {
+                chain_id: TestNode::<()>::CHAIN_ID.to_string(),
+                ..Default::default()
+            },
+        };
+        plan.populate_detection_data(rand_core::OsRng, 0);
+        (plan, undelegate_token_id)
+    };
+    let tx = client.witness_auth_build(&plan).await?;
+
+    // Execute the undelegation transaction, applying it to the chain state.
+    let pre_undelegated_epoch = get_latest_epoch().await;
+    node.block()
+        .add_tx(tx.encode_to_vec())
+        .execute()
+        .instrument(error_span!("executing block with undelegation transaction"))
+        .await?;
+    let post_undelegate_snapshot = storage.latest_snapshot();
+
+    // Compute the height we expect to see this unbonding period finish.
+    let expected_end_of_unboding_period_height = {
+        // TODO(kate): right now the unbonding delay is calculated by using the start of the epoch
+        // as a beginning. so rather than...
+        // post_undelegated_height + UNBONDING_DELAY
+        // ...we write:
+        pre_undelegated_epoch.start_height + UNBONDING_DELAY
```

Comment:
There's a `compute_unbonding_height` routine maybe we should create a stateless version of it that we can reuse in tests and other places? This works too

Reply:
yeah! that's a good idea. i waffled :waffle: on whether or not to compute this here or use that method, but you're right that this is probably easier. fixed in d938eae!